WalletWasabi/WabiSabi/Models/Serialization/GenericInterfaceJsonConverter.cs

```diff
@@ -0,0 +1,68 @@
+using System.Collections.Generic;
+using System.Linq;
+using Newtonsoft.Json;
+using Newtonsoft.Json.Linq;
+
+namespace WalletWasabi.WabiSabi.Models.Serialization;
+
+public class GenericInterfaceJsonConverter<T> : JsonConverter<T>
+{
+	// This converter is a bit unusual because we need to add a new property to the
+	// serialized json string but the converter is called recursively and fails with
+	// an "Self referencing loop" exception.
+	// The workaround is detect it and prevent reentering by setting CanRead and
+	// CanWrite to false immediately after entering.
+	// see: https://github.com/JamesNK/Newtonsoft.Json/issues/386
+	[ThreadStatic]
+	private static bool IsReading;
+
+	[ThreadStatic]
+	private static bool IsWriting;
+
+	public GenericInterfaceJsonConverter(IEnumerable<Type> types)
+	{
+		Types = types;
```

Comment:
Imho we should convert here `types` into a `Dictionary<string, Type>`. It's more efficient and seems even more straightforward.

Reply:
I would write it like this: https://github.com/kiminuo/WalletWasabi/commit/59c1040454f5514c39d80fc80493c5d2ce941829 (https://github.com/kiminuo/WalletWasabi/tree/feature/2022-02-03-rewrite-generic-serializer) and add a few tests (I couldn't find existing tests).

Reply:
There are tests for serialization and deserialization of messages that use this new generic converter but i agree they are not enough. I decided to push the PR "as is" (with only a few UTs) to reduce the uncertainty around this feature but I really would like to have tests with real request/responses to verify the polymorphism cases are correct instead of testing with other types like `Person`.

About the dictionary, well, yes, it can be but we have 2 instances in one case and 3 in the largest case so no performance is gained.